File: test/calendar-reminders.test.ts
import {
  addMember,
  api,
  createUser,
  db,
  rig,
  services,
  setPlan,
  signIn,
} from "./helpers";
import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
} from "bun:test";
import { eq } from "drizzle-orm";
import { schema } from "../src/server/db";
import { clockFmt, runCalendarReminders } from "../src/server/scheduled";

const SUB_KEYS = {
  p256dh:
    "BNcRdreALRFXTkOOUHK1EtK2wtaz5Ry4YfYCA_0QTpQtUbVlUls0VJXg7A8u-Ts1XbjhazAkj7I99e8QcYP7DkM",
  auth: "tBHItJI5svbpez7KI4CCXg",
};
const PUSH_ORIGIN = "https://fcm.googleapis.com";
const HOUR = 3600_000;

const delivered: string[] = [];
const realFetch = globalThis.fetch;

beforeAll(() => {
  globalThis.fetch = (async (
    input: Parameters<typeof fetch>[0],
    init?: RequestInit,
  ) => {
    const url =
      typeof input === "string"
        ? input
        : input instanceof URL
          ? input.href
          : input.url;
    if (url.startsWith(PUSH_ORIGIN)) {
      delivered.push(url);
      return new Response("", { status: 201 });
    }
    return realFetch(input as never, init);
  }) as typeof fetch;
});

afterAll(() => {
  globalThis.fetch = realFetch;
});

beforeEach(() => {
  delivered.length = 0;
});

async function subscribe(cookie: string, endpoint: string) {
  const res = await api("/api/push/subscribe", {
    method: "POST",
    cookie,
    body: { endpoint, ...SUB_KEYS },
  });
  expect(res.status).toBe(200);
}

async function createEvent(
  cookie: string,
  body: Record<string, unknown>,
): Promise<string> {
  const res = await api("/api/calendar/events", {
    method: "POST",
    cookie,
    body,
  });
  expect(res.status).toBe(201);
  return ((await res.json()) as { id: string }).id;
}

describe("calendar reminders", () => {
  it("fires once inside the window, to all members when unassigned", async () => {
    const a = await rig();
    await setPlan(a.family.id, "premium");
    await subscribe(a.cookie, `${PUSH_ORIGIN}/cal/admin`);
    const other = await createUser("Other parent");
    await addMember(other.id, a.family.id, "member");
    const otherCookie = await signIn(other.email);
    await subscribe(otherCookie, `${PUSH_ORIGIN}/cal/other`);

    const now = Date.now();
    await createEvent(a.cookie, {
      title: "Checkup",
      startTime: new Date(now + 30 * 60_000).toISOString(),
      remindMinutesBefore: 60,
    });

    // Inside the lead window: one push per member, then silence.
    expect(await runCalendarReminders(services, now)).toBe(2);
    expect(delivered.sort()).toEqual([
      `${PUSH_ORIGIN}/cal/admin`,
      `${PUSH_ORIGIN}/cal/other`,
    ]);
    expect(await runCalendarReminders(services, now + 15 * 60_000)).toBe(0);
  });

  it("targets only assignees when set", async () => {
    const a = await rig();
    await setPlan(a.family.id, "premium");
    await subscribe(a.cookie, `${PUSH_ORIGIN}/cal2/admin`);
    const other = await createUser("Other parent");
    await addMember(other.id, a.family.id, "member");
    const otherCookie = await signIn(other.email);
    await subscribe(otherCookie, `${PUSH_ORIGIN}/cal2/other`);

    const now = Date.now();
    await createEvent(a.cookie, {
      title: "Babysitting",
      startTime: new Date(now + 30 * 60_000).toISOString(),
      remindMinutesBefore: 60,
      assigneeUserIds: [other.id],
    });
    expect(await runCalendarReminders(services, now)).toBe(1);
    expect(delivered).toEqual([`${PUSH_ORIGIN}/cal2/other`]);
  });

  it("not yet due → nothing; long-past → latched silently", async () => {
    const a = await rig();
    await setPlan(a.family.id, "premium");
    await subscribe(a.cookie, `${PUSH_ORIGIN}/cal3/admin`);

    const now = Date.now();
    const notDue = await createEvent(a.cookie, {
      title: "Far future",
      startTime: new Date(now + 10 * HOUR).toISOString(),
      remindMinutesBefore: 60,
    });
    const past = await createEvent(a.cookie, {
      title: "Missed",
      startTime: new Date(now + HOUR).toISOString(),
      remindMinutesBefore: 60,
    });
    // Simulate downtime: the sweep first runs 2h after the past event started.
    expect(await runCalendarReminders(services, now + 3 * HOUR)).toBe(0);
    expect(delivered).toHaveLength(0);

    const rows = await db()
      .select({
        id: schema.calendarEvent.id,
        remindedAt: schema.calendarEvent.remindedAt,
      })
      .from(schema.calendarEvent)
      .where(eq(schema.calendarEvent.familyId, a.family.id));
    expect(rows.find((r) => r.id === past)!.remindedAt).not.toBeNull();
    expect(rows.find((r) => r.id === notDue)!.remindedAt).toBeNull();
  });

  it("formats the reminder clock in Europe/Oslo, not workerd's UTC default", async () => {
    // The pushed body itself isn't observable through the fetch stub above —
    // web-push encrypts the JSON payload (aes128gcm) before the HTTP call,
    // so the stub only ever sees ciphertext bytes. clockFmt is exported from
    // scheduled.ts specifically so this regression (a 14:00 CEST appointment
    // rendering as "12:00") can be asserted directly at the unit level.
    // 2026-08-25T12:00:00Z is during CEST (UTC+2) → 14:00 Oslo-local.
    expect(clockFmt.format(new Date("2026-08-25T12:00:00Z"))).toBe("14:00");
    // Winter (CET, UTC+1) sanity check too.
    expect(clockFmt.format(new Date("2026-01-15T12:00:00Z"))).toBe("13:00");

    // And a full round trip still delivers exactly once per due event, as
    // the other tests in this file already assert in detail.
    const a = await rig();
    await setPlan(a.family.id, "premium");
    await subscribe(a.cookie, `${PUSH_ORIGIN}/cal4/admin`);
    const now = Date.now();
    await createEvent(a.cookie, {
      title: "Oslo clock check",
      startTime: new Date(now + 30 * 60_000).toISOString(),
      remindMinutesBefore: 60,
    });
    expect(await runCalendarReminders(services, now)).toBe(1);
  });
});

File: apps/server/internal/jobs/calendar_reminders_test.go
package jobs_test

import (
	"context"
	"net/http"
	"testing"
	"time"

	"github.com/refsdal/pjokk/server/internal/auth"
	"github.com/refsdal/pjokk/server/internal/jobs"
	"github.com/refsdal/pjokk/server/internal/testrig"
)

const (
	calSubP256dh = "BNcRdreALRFXTkOOUHK1EtK2wtaz5Ry4YfYCA_0QTpQtUbVlUls0VJXg7A8u-Ts1XbjhazAkj7I99e8QcYP7DkM"
	calSubAuth   = "tBHItJI5svbpez7KI4CCXg"
)

func subscribePush(t *testing.T, a *testrig.AppRig, cookie, endpoint string) {
	t.Helper()
	res := a.Do(http.MethodPost, "/api/push/subscribe", cookie, map[string]any{
		"endpoint": endpoint,
		"p256dh":   calSubP256dh,
		"auth":     calSubAuth,
	})
	if res.Status != http.StatusOK {
		t.Fatalf("subscribe %q status = %d, body %s", endpoint, res.Status, res.Raw)
	}
}

func createCalendarEvent(t *testing.T, a *testrig.AppRig, cookie string, body map[string]any) string {
	t.Helper()
	res := a.Do(http.MethodPost, "/api/calendar/events", cookie, body)
	if res.Status != http.StatusCreated {
		t.Fatalf("create event status = %d, body %s", res.Status, res.Raw)
	}
	id, _ := res.JSON["id"].(string)
	if id == "" {
		t.Fatalf("create event: no id in response %s", res.Raw)
	}
	return id
}

func userIDByEmail(t *testing.T, a *testrig.AppRig, email string) string {
	t.Helper()
	var id string
	if err := a.Rig.Pool.QueryRow(context.Background(),
		`SELECT "id" FROM "users" WHERE "email" = $1`, email).Scan(&id); err != nil {
		t.Fatalf("userIDByEmail(%q): %v", email, err)
	}
	return id
}

// Ports calendar-reminders.test.ts's "fires once inside the window, to all
// members when unassigned".
func TestCalendarRemindersFireOnceToAllMembersWhenUnassigned(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	adminID := userIDByEmail(t, a, "parent@example.com")
	subscribePush(t, a, cookie, "https://fcm.googleapis.com/cal/admin")

	otherID := a.SignUp("Other parent", "other@example.com")
	otherCookie := a.AddMember(familyID, otherID, auth.RoleMember, "other@example.com")
	subscribePush(t, a, otherCookie, "https://fcm.googleapis.com/cal/other")

	now := time.Now().Truncate(time.Second)
	createCalendarEvent(t, a, cookie, map[string]any{
		"title":               "Checkup",
		"startTime":           now.Add(30 * time.Minute).Format(time.RFC3339),
		"remindMinutesBefore": 60,
	})

	d := depsFor(a)
	ctx := context.Background()

	// Inside the lead window: one push per member, then silence.
	sent, err := jobs.RunCalendarReminders(ctx, d, now)
	if err != nil {
		t.Fatalf("RunCalendarReminders: %v", err)
	}
	if sent != 2 {
		t.Fatalf("sent = %d, want 2", sent)
	}
	if got := a.Push.Count(adminID); got != 1 {
		t.Errorf("admin delivery count = %d, want 1", got)
	}
	if got := a.Push.Count(otherID); got != 1 {
		t.Errorf("other member delivery count = %d, want 1", got)
	}

	sentAgain, err := jobs.RunCalendarReminders(ctx, d, now.Add(15*time.Minute))
	if err != nil {
		t.Fatalf("RunCalendarReminders (repeat): %v", err)
	}
	if sentAgain != 0 {
		t.Fatalf("sent (repeat) = %d, want 0", sentAgain)
	}
}

// Ports "targets only assignees when set".
func TestCalendarRemindersTargetOnlyAssigneesWhenSet(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	adminID := userIDByEmail(t, a, "parent@example.com")
	subscribePush(t, a, cookie, "https://fcm.googleapis.com/cal2/admin")

	otherID := a.SignUp("Other parent", "other2@example.com")
	otherCookie := a.AddMember(familyID, otherID, auth.RoleMember, "other2@example.com")
	subscribePush(t, a, otherCookie, "https://fcm.googleapis.com/cal2/other")

	now := time.Now().Truncate(time.Second)
	createCalendarEvent(t, a, cookie, map[string]any{
		"title":               "Babysitting",
		"startTime":           now.Add(30 * time.Minute).Format(time.RFC3339),
		"remindMinutesBefore": 60,
		"assigneeUserIds":     []string{otherID},
	})

	d := depsFor(a)
	sent, err := jobs.RunCalendarReminders(context.Background(), d, now)
	if err != nil {
		t.Fatalf("RunCalendarReminders: %v", err)
	}
	if sent != 1 {
		t.Fatalf("sent = %d, want 1", sent)
	}
	if got := a.Push.Count(otherID); got != 1 {
		t.Errorf("assignee delivery count = %d, want 1", got)
	}
	if got := a.Push.Count(adminID); got != 0 {
		t.Errorf("non-assignee delivery count = %d, want 0", got)
	}
}

// Ports "not yet due -> nothing; long-past -> latched silently".
func TestCalendarRemindersGraceLatchesLongPastEventsSilently(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	adminID := userIDByEmail(t, a, "parent@example.com")
	subscribePush(t, a, cookie, "https://fcm.googleapis.com/cal3/admin")

	now := time.Now().Truncate(time.Second)
	const hour = time.Hour

	notDueID := createCalendarEvent(t, a, cookie, map[string]any{
		"title":               "Far future",
		"startTime":           now.Add(10 * hour).Format(time.RFC3339),
		"remindMinutesBefore": 60,
	})
	pastID := createCalendarEvent(t, a, cookie, map[string]any{
		"title":               "Missed",
		"startTime":           now.Add(hour).Format(time.RFC3339),
		"remindMinutesBefore": 60,
	})

	d := depsFor(a)
	ctx := context.Background()

	// Simulate downtime: the sweep first runs 2h after the past event started.
	sent, err := jobs.RunCalendarReminders(ctx, d, now.Add(3*hour))
	if err != nil {
		t.Fatalf("RunCalendarReminders: %v", err)
	}
	if sent != 0 {
		t.Fatalf("sent = %d, want 0", sent)
	}
	if got := a.Push.Count(adminID); got != 0 {
		t.Errorf("delivery count = %d, want 0 (latched silently)", got)
	}

	remindedAt := calendarEventRemindedAt(t, a, pastID)
	if remindedAt == nil {
		t.Errorf("past event's reminded_at is nil, want latched")
	}
	notDueRemindedAt := calendarEventRemindedAt(t, a, notDueID)
	if notDueRemindedAt != nil {
		t.Errorf("not-due event's reminded_at = %v, want nil", *notDueRemindedAt)
	}
}

func calendarEventRemindedAt(t *testing.T, a *testrig.AppRig, id string) *time.Time {
	t.Helper()
	var remindedAt *time.Time
	if err := a.Rig.Pool.QueryRow(context.Background(),
		`SELECT "reminded_at" FROM "calendar_event" WHERE "id" = $1`, id).Scan(&remindedAt); err != nil {
		t.Fatalf("read reminded_at for %q: %v", id, err)
	}
	return remindedAt
}

// Ports "formats the reminder clock in Europe/Oslo, not workerd's UTC
// default" — the direct clockFmt assertions, plus one round trip.
func TestFormatOsloClock(t *testing.T) {
	// 2026-08-25T12:00:00Z is during CEST (UTC+2) -> 14:00 Oslo-local.
	summer := time.Date(2026, 8, 25, 12, 0, 0, 0, time.UTC)
	if got := jobs.FormatOsloClock(summer); got != "14:00" {
		t.Errorf("FormatOsloClock(%v) = %q, want %q", summer, got, "14:00")
	}

	// Winter (CET, UTC+1) sanity check too.
	winter := time.Date(2026, 1, 15, 12, 0, 0, 0, time.UTC)
	if got := jobs.FormatOsloClock(winter); got != "13:00" {
		t.Errorf("FormatOsloClock(%v) = %q, want %q", winter, got, "13:00")
	}
}

func TestCalendarRemindersRoundTripDeliversExactlyOncePerDueEvent(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	subscribePush(t, a, cookie, "https://fcm.googleapis.com/cal4/admin")

	now := time.Now().Truncate(time.Second)
	createCalendarEvent(t, a, cookie, map[string]any{
		"title":               "Oslo clock check",
		"startTime":           now.Add(30 * time.Minute).Format(time.RFC3339),
		"remindMinutesBefore": 60,
	})

	d := depsFor(a)
	sent, err := jobs.RunCalendarReminders(context.Background(), d, now)
	if err != nil {
		t.Fatalf("RunCalendarReminders: %v", err)
	}
	if sent != 1 {
		t.Fatalf("sent = %d, want 1", sent)
	}
}
